pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultipleOrEqualitiesToInClauseFilterQueryTreeOptimizer.java

```diff
@@ -18,163 +18,221 @@
+  private FilterQueryTree singlePredicateOptimization(FilterQueryTree filterQuery) {
+    // rewrite single predicate IN as EQUALITY, for remaining cases this optimizer is a NO-OP
+    if (filterQuery.getOperator() == FilterOperator.IN && filterQuery.getValue().size() == 1) {
+      filterQuery.setOperator(FilterOperator.EQUALITY);
+    }
+    return filterQuery;
+  }
 
-      // If we have at least one column to return
-      if (!columnToValues.isEmpty()) {
-        // We can eliminate the OR node if there is only one column with multiple values
-        if (columnToValues.size() == 1 && nonEqualityOperators.isEmpty()) {
-          Map.Entry<String, Set<String>> columnAndValues = columnToValues.entrySet().iterator().next();
+  private static class PredicateInfo {
+    Set<String> _predicateValues;
+    boolean _multipleOccurences;
+    boolean _inPredicate;
 
-          return buildFilterQueryTreeForColumnAndValues(columnAndValues);
-        }
+    PredicateInfo(Set<String> predicateValues, boolean multipleOccurences, boolean inPredicate) {
+      _predicateValues = predicateValues;
+      _multipleOccurences = multipleOccurences;
+      _inPredicate = inPredicate;
+    }
+  }
 
-        // Check if we need to rebuild the predicate
-        boolean rebuildRequired = isRebuildRequired(columnToValues, containsDuplicates);
+  private FilterQueryTree optimizeTreeRootedAtOR(FilterQueryTree filterQueryTree) {
+    Map<String, PredicateInfo> columnToInEqPredicateInfo = new HashMap<>();
+    Map<FilterQueryTree, Integer> nonEqInOperators = new HashMap<>();
+
+    // Collect all equality/in operators and non-equality operators
+    // these are the immediate children of root OR operator
+    boolean containsDuplicates =
+        collectChildOperatorsOfRootOROperator(filterQueryTree, columnToInEqPredicateInfo, nonEqInOperators);
+
+    if (columnToInEqPredicateInfo.size() == 1 && nonEqInOperators.isEmpty()) {
+      // We can eliminate the OR root node if there is exactly one unique column with one or more
+      // predicate values(could be duplicate) using IN/OR. Rewritten as follows:
+      // -- if there is one predicate (after deduping), use EQUALITY
+      // -- else use IN
+      // e.g:
+      // a = 1 OR a = 1 --> can be rewritten as a = 1
+      // a = 1 OR a = 2 OR a = 3 --> can be rewritten as a IN (1, 2, 3)
+      Map.Entry<String, PredicateInfo> columnAndValues = columnToInEqPredicateInfo.entrySet().iterator().next();
+      return buildFilterQueryTreeForColumnAndPredicateInfo(columnAndValues.getKey(), columnAndValues.getValue());
+    }
 
-        if (!rebuildRequired) {
-          // No mutation needed, so just return the same tree
-          return filterQueryTree;
-        } else {
-          // Rebuild the predicates
-          return rebuildFilterPredicate(columnToValues, nonEqualityOperators);
+    boolean rewriteRequiredForInEqOperators = containsDuplicates || isRebuildRequiredForInEqPredicates(columnToInEqPredicateInfo);
```

Comment:
The logic is more or less similar to old but some additional work was needed to handle nested better. Secondly, the new code handles the no-rewrite cases better.